File: packages/agent/src/core/utils/message-sanitizer.ts
/**
 * @fileoverview Message Sanitizer - Single Source of Truth for API Compliance
 *
 * This module provides the definitive sanitization layer for messages before
 * they are sent to any LLM provider API. It guarantees that the output will
 * be valid for the Anthropic/OpenAI/Google APIs.
 *
 * Properties:
 * - Pure: No side effects, deterministic output
 * - Idempotent: sanitize(sanitize(x)) === sanitize(x)
 * - Total: Handles any input, never throws
 * - Documented: Returns list of all fixes applied
 *
 * Invariants enforced:
 * 1. Every tool_use block has a corresponding toolResult message
 * 2. No empty messages (empty content)
 * 3. No toolResult without valid toolCallId
 * 4. First message is user role (after conversion to API format)
 */

import type {
  Message,
  UserMessage,
  AssistantMessage,
  ToolResultMessage,
  AssistantContent,
  ToolCall,
} from '@core/types/messages.js';
import { isToolCall } from '@core/types/messages.js';
import { createLogger } from '@infrastructure/logging/index.js';

const logger = createLogger('message-sanitizer');

// =============================================================================
// Types
// =============================================================================

export type SanitizationFixType =
  | 'injected_tool_result'
  | 'removed_empty_message'
  | 'removed_thinking_only_message'
  | 'removed_invalid_block'
  | 'removed_duplicate_tool_use'
  | 'merged_consecutive_messages'
  | 'injected_placeholder_user';

export interface SanitizationFix {
  type: SanitizationFixType;
  details: Record<string, unknown>;
}

export interface SanitizationResult {
  /** Sanitized messages array - guaranteed valid for API submission */
  messages: Message[];
  /** List of fixes that were applied */
  fixes: SanitizationFix[];
  /** True if no fixes were needed (input was already valid) */
  isValid: boolean;
}

export type ValidationViolationType =
  | 'missing_tool_result'
  | 'empty_message'
  | 'invalid_tool_result'
  | 'missing_first_user';

export interface ValidationViolation {
  type: ValidationViolationType;
  details: Record<string, unknown>;
}

// =============================================================================
// Constants
// =============================================================================

/** Content for synthetic tool results when execution was interrupted */
const INTERRUPTED_CONTENT = '[Interrupted]';

/** Content for placeholder user message when conversation doesn't start with user */
const PLACEHOLDER_USER_CONTENT = '[Continued]';

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Deep clone a message to avoid mutating input.
 * Uses JSON serialization for simplicity and to ensure clean serializable output.
 */
function cloneMessage<T extends Message>(msg: T): T {
  try {
    return JSON.parse(JSON.stringify(msg));
  } catch {
    // If serialization fails, return a minimal valid structure
    return { ...msg } as T;
  }
}

/**
 * Check if a user message has valid non-empty content.
 */
function isValidUserMessage(msg: UserMessage): boolean {
  if (typeof msg.content === 'string') {
    return msg.content.trim().length > 0;
  }
  if (Array.isArray(msg.content)) {
    return msg.content.length > 0;
  }
  return false;
}

/**
 * Check if an assistant message has content that will survive API conversion.
 *
 * Thinking blocks without signatures are display-only (used by non-extended thinking
 * models like Haiku/Sonnet) and are filtered out before sending to the API.
 * If a message contains ONLY such thinking blocks, it will become empty after conversion
 * and must be removed to prevent API errors.
 *
 * Content that survives conversion:
 * - text blocks (always)
 * - tool_use blocks (always)
 * - thinking blocks WITH signatures (extended thinking models like Opus 4.5)
 */
function hasContentSurvivingConversion(content: AssistantContent[]): boolean {
  return content.some(block => {
    if (!block || typeof block !== 'object') return false;
    const type = block.type;

    // text and tool_use blocks always survive
    if (type === 'text' || type === 'tool_use') return true;

    // thinking blocks only survive if they have a signature (extended thinking)
    if (type === 'thinking') {
      return typeof (block as { signature?: string }).signature === 'string' &&
             (block as { signature?: string }).signature!.length > 0;
    }

    // Unknown block types - be permissive and keep them
    return true;
  });
}

/**
 * Check if an assistant message has valid non-empty content.
 * Also checks that the content will survive API conversion (thinking-only messages
 * without signatures will become empty and are therefore invalid).
 */
function isValidAssistantMessage(msg: AssistantMessage): boolean {
  if (!Array.isArray(msg.content)) {
    return false;
  }
  if (msg.content.length === 0) {
    return false;
  }
  // Must have content that survives API conversion
  return hasContentSurvivingConversion(msg.content);
}

/**
 * Check if a tool result message has valid required fields.
 */
function isValidToolResultMessage(msg: ToolResultMessage): boolean {
  return typeof msg.toolCallId === 'string' && msg.toolCallId.length > 0;
}

/**
 * Check if a message is valid (non-empty and well-formed).
 */
function isValidMessage(msg: Message): boolean {
  if (!msg || typeof msg !== 'object' || !msg.role) {
    return false;
  }

  switch (msg.role) {
    case 'user':
      return isValidUserMessage(msg as UserMessage);
    case 'assistant':
      return isValidAssistantMessage(msg as AssistantMessage);
    case 'toolResult':
      return isValidToolResultMessage(msg as ToolResultMessage);
    default:
      return false;
  }
}

/**
 * Extract all tool_use IDs from an assistant message.
 */
function extractToolUseIds(msg: AssistantMessage): string[] {
  if (!Array.isArray(msg.content)) {
    return [];
  }

  return msg.content
    .filter((block): block is ToolCall =>
      block != null &&
      isToolCall(block) &&
      block.id.length > 0
    )
    .map(block => block.id);
}

// =============================================================================
// Main Functions
// =============================================================================

/**
 * Sanitize messages to guarantee API compliance.
 *
 * This is the SINGLE SOURCE OF TRUTH for message validity.
 * It can accept ANY input (including completely malformed) and
 * will always produce a valid Message[] that can be converted
 * to any provider format.
 *
 * @param messages - Input messages (may be malformed)
 * @returns Sanitized messages with list of fixes applied
 */
export function sanitizeMessages(messages: Message[]): SanitizationResult {
  const fixes: SanitizationFix[] = [];

  // Handle invalid input gracefully
  if (!Array.isArray(messages)) {
    logger.warn('sanitizeMessages received non-array input', { type: typeof messages });
    return { messages: [], fixes: [], isValid: true };
  }

  // PHASE 1: Filter invalid messages, deduplicate tool_use IDs, and build valid message list
  const validMessages: Message[] = [];
  const toolUseLocations = new Map<string, number>(); // toolUseId → index in validMessages where assistant is
  const seenToolUseIds = new Set<string>(); // For deduplication across messages

  for (const msg of messages) {
    if (!isValidMessage(msg)) {
      // Determine the specific reason for removal
      let fixType: SanitizationFixType = 'removed_empty_message';

      // Check if this is a thinking-only assistant message
      if (msg?.role === 'assistant' && Array.isArray((msg as AssistantMessage).content)) {
        const content = (msg as AssistantMessage).content;
        if (content.length > 0 && !hasContentSurvivingConversion(content)) {
          fixType = 'removed_thinking_only_message';
          logger.warn('Removed assistant message with only thinking blocks (no signature)', {
            blockCount: content.length,
            blockTypes: content.map((b: AssistantContent) => b?.type),
          });
        }
      }

      fixes.push({
        type: fixType,
        details: { role: msg?.role ?? 'unknown' },
      });
      continue;
    }

    const cloned = cloneMessage(msg);

    // Deduplicate tool_use blocks within assistant messages
    if (cloned.role === 'assistant') {
      const assistantMsg = cloned as AssistantMessage;
      const originalLength = assistantMsg.content.length;
      assistantMsg.content = assistantMsg.content.filter(block => {
        if (isToolCall(block)) {
          if (seenToolUseIds.has(block.id)) {
            fixes.push({
              type: 'removed_duplicate_tool_use',
              details: { toolUseId: block.id },
            });
            logger.warn('Removed duplicate tool_use block', { toolUseId: block.id });
            return false;
          }
          seenToolUseIds.add(block.id);
        }
        return true;
      });

      // If message became empty after dedup, skip it
      if (!isValidAssistantMessage(assistantMsg)) {
        if (originalLength > 0) {
          fixes.push({
            type: 'removed_empty_message',
            details: { role: 'assistant', reason: 'empty_after_dedup' },
          });
        }
        continue;
      }
    }

    const index = validMessages.length;
    validMessages.push(cloned);

    // Track tool_use locations for later injection
    if (cloned.role === 'assistant') {
      const toolUseIds = extractToolUseIds(cloned as AssistantMessage);
      for (const id of toolUseIds) {
        toolUseLocations.set(id, index);
      }
    }
  }

  // PHASE 2: Collect existing tool result IDs
  const existingToolResultIds = new Set<string>();
  for (const msg of validMessages) {
    if (msg.role === 'toolResult') {
      existingToolResultIds.add((msg as ToolResultMessage).toolCallId);
    }
  }

  // PHASE 3: Find missing tool results and inject synthetic ones
  // We need to inject after each assistant message that has unmatched tool_use blocks
  // Work backwards to maintain correct indices during insertion
  const missingByAssistantIndex = new Map<number, string[]>();

  for (const [toolUseId, assistantIndex] of toolUseLocations) {
    if (!existingToolResultIds.has(toolUseId)) {
      const existing = missingByAssistantIndex.get(assistantIndex) || [];
      existing.push(toolUseId);
      missingByAssistantIndex.set(assistantIndex, existing);
    }
  }

  // Sort by assistant index descending so we can insert without shifting issues
  const sortedIndices = [...missingByAssistantIndex.keys()].sort((a, b) => b - a);

  for (const assistantIndex of sortedIndices) {
    const missingIds = missingByAssistantIndex.get(assistantIndex)!;

    // Insert synthetic tool results immediately after the assistant message
    // Insert in reverse order of the missing IDs to maintain original tool_use order
    for (let i = missingIds.length - 1; i >= 0; i--) {
      const toolCallId = missingIds[i]!;
      const syntheticResult: ToolResultMessage = {
        role: 'toolResult',
        toolCallId,
        content: INTERRUPTED_CONTENT,
        isError: false,
      };

      validMessages.splice(assistantIndex + 1, 0, syntheticResult);

      fixes.push({
        type: 'injected_tool_result',
        details: { toolCallId, afterIndex: assistantIndex },
      });

      logger.warn('Injected synthetic tool result for interrupted execution', { toolCallId });
    }
  }

  // PHASE 4: Ensure first message is user (for API compliance after conversion)
  if (validMessages.length > 0 && validMessages[0]!.role !== 'user') {
    validMessages.unshift({
      role: 'user',
      content: PLACEHOLDER_USER_CONTENT,
    });

    fixes.push({
      type: 'injected_placeholder_user',
      details: {},
    });

    logger.warn('Injected placeholder user message at start');
  }

  // PHASE 5: Merge consecutive same-role messages (except toolResult)
  // This catches any source of consecutive same-role messages, including
  // those created by removing thinking-only messages or dedup operations.
  const mergedMessages: Message[] = [];
  for (const msg of validMessages) {
    const prev = mergedMessages[mergedMessages.length - 1];
    if (prev && prev.role === msg.role && msg.role !== 'toolResult') {
      if (msg.role === 'assistant') {
        // Merge assistant: concatenate content arrays
        const prevAssistant = prev as AssistantMessage;
        const curAssistant = msg as AssistantMessage;
        prevAssistant.content = [...prevAssistant.content, ...curAssistant.content];
      } else if (msg.role === 'user') {
        // Merge user: normalize both to array format then concatenate
        const prevUser = prev as UserMessage;
        const curUser = msg as UserMessage;
        const prevContent = typeof prevUser.content === 'string'
          ? [{ type: 'text' as const, text: prevUser.content }]
          : Array.isArray(prevUser.content) ? prevUser.content : [];
        const curContent = typeof curUser.content === 'string'
          ? [{ type: 'text' as const, text: curUser.content }]
          : Array.isArray(curUser.content) ? curUser.content : [];
        prevUser.content = [...prevContent, ...curContent] as UserMessage['content'];
      }
      fixes.push({
        type: 'merged_consecutive_messages',
        details: { role: msg.role },
      });
      logger.warn('Merged consecutive same-role messages', { role: msg.role });
    } else {
      mergedMessages.push(msg);
    }
  }

  return {
    messages: mergedMessages,
    fixes,
    isValid: fixes.length === 0,
  };
}

/**
 * Validate messages without fixing.
 * Returns list of violations found.
 *
 * Use this for diagnostics/logging without modifying messages.
 *
 * @param messages - Messages to validate
 * @returns List of validation violations
 */
export function validateMessages(messages: Message[]): ValidationViolation[] {
  const violations: ValidationViolation[] = [];

  if (!Array.isArray(messages)) {
    return violations;
  }

  // Collect tool_use IDs and existing tool result IDs
  const toolUseIds = new Set<string>();
  const toolResultIds = new Set<string>();

  for (const msg of messages) {
    if (!isValidMessage(msg)) {
      violations.push({
        type: 'empty_message',
        details: { role: msg?.role ?? 'unknown' },
      });
      continue;
    }

    if (msg.role === 'assistant') {
      const ids = extractToolUseIds(msg as AssistantMessage);
      for (const id of ids) {
        toolUseIds.add(id);
      }
    }

    if (msg.role === 'toolResult') {
      const trMsg = msg as ToolResultMessage;
      if (!trMsg.toolCallId) {
        violations.push({
          type: 'invalid_tool_result',
          details: { reason: 'missing toolCallId' },
        });
      } else {
        toolResultIds.add(trMsg.toolCallId);
      }
    }
  }

  // Check for missing tool results
  for (const id of toolUseIds) {
    if (!toolResultIds.has(id)) {
      violations.push({
        type: 'missing_tool_result',
        details: { toolCallId: id },
      });
    }
  }

  // Check first message is user
  const validMessages = messages.filter(isValidMessage);
  if (validMessages.length > 0 && validMessages[0]!.role !== 'user') {
    violations.push({
      type: 'missing_first_user',
      details: { firstRole: validMessages[0]!.role },
    });
  }

  return violations;
}

File: packages/agent-rs/crates/tron-llm-anthropic/src/message_sanitizer.rs
//! # Message Sanitizer
//!
//! Pre-conversion sanitization for messages before they're sent to the Anthropic API.
//! Ports phases 1–4 from the TypeScript `sanitizeMessages()`.
//!
//! Invariants enforced:
//! 1. Every `tool_use` block has a corresponding `ToolResult` message
//! 2. No empty messages
//! 3. No thinking-only assistant messages (display-only, no signature)
//! 4. First message is user role
//!
//! This is idempotent: `sanitize(sanitize(x)) == sanitize(x)`.

use std::collections::{HashMap, HashSet};

use tracing::warn;

use tron_core::content::AssistantContent;
use tron_core::messages::{Message, ToolResultMessageContent};

/// Content for synthetic tool results when execution was interrupted.
const INTERRUPTED_CONTENT: &str = "[Interrupted]";

/// Content for placeholder user message when conversation doesn't start with user.
const CONTINUED_CONTENT: &str = "[Continued]";

/// Sanitize messages to guarantee API compliance.
///
/// Returns a new `Vec<Message>` with fixes applied:
/// - Empty messages filtered out
/// - Thinking-only assistant messages (no signature) filtered out
/// - Duplicate tool_use IDs deduplicated
/// - Synthetic tool results injected for unmatched tool_use blocks
/// - Placeholder user message prepended if first message isn't user
pub fn sanitize_messages(messages: Vec<Message>) -> Vec<Message> {
    // PHASE 1: Filter invalid messages, deduplicate tool_use IDs
    let mut valid: Vec<Message> = Vec::with_capacity(messages.len());
    let mut seen_tool_use_ids: HashSet<String> = HashSet::new();
    // tool_use_id → index in `valid` where the assistant message lives
    let mut tool_use_locations: HashMap<String, usize> = HashMap::new();

    for msg in messages {
        match &msg {
            Message::User { content, .. } => {
                if !is_valid_user_content(content) {
                    warn!("Removed empty user message");
                    continue;
                }
                valid.push(msg);
            }
            Message::Assistant { content, .. } => {
                // Filter out duplicate tool_use blocks
                let mut filtered: Vec<AssistantContent> = Vec::with_capacity(content.len());
                for block in content {
                    if let AssistantContent::ToolUse { id, .. } = block {
                        if seen_tool_use_ids.contains(id) {
                            warn!(tool_use_id = %id, "Removed duplicate tool_use block");
                            continue;
                        }
                        let _ = seen_tool_use_ids.insert(id.clone());
                    }
                    filtered.push(block.clone());
                }

                // Check if remaining content survives API conversion
                if !has_content_surviving_conversion(&filtered) {
                    warn!("Removed assistant message with no content surviving conversion");
                    continue;
                }

                let idx = valid.len();

                // Track tool_use locations for synthetic result injection
                for block in &filtered {
                    if let AssistantContent::ToolUse { id, .. } = block {
                        let _ = tool_use_locations.insert(id.clone(), idx);
                    }
                }

                valid.push(Message::Assistant {
                    content: filtered,
                    usage: None,
                    cost: None,
                    stop_reason: None,
                    thinking: None,
                });
            }
            Message::ToolResult {
                tool_call_id,
                content,
                ..
            } => {
                if tool_call_id.is_empty() {
                    warn!("Removed tool result with empty tool_call_id");
                    continue;
                }
                if !is_valid_tool_result_content(content) {
                    warn!(tool_call_id = %tool_call_id, "Removed empty tool result");
                    continue;
                }
                valid.push(msg);
            }
        }
    }

    // PHASE 2: Collect existing tool result IDs
    let existing_result_ids: HashSet<&str> = valid
        .iter()
        .filter_map(|msg| {
            if let Message::ToolResult { tool_call_id, .. } = msg {
                Some(tool_call_id.as_str())
            } else {
                None
            }
        })
        .collect();

    // PHASE 3: Inject synthetic tool results for unmatched tool_use blocks
    // Group missing IDs by assistant message index
    let mut missing_by_index: HashMap<usize, Vec<String>> = HashMap::new();
    for (tool_use_id, assistant_idx) in &tool_use_locations {
        if !existing_result_ids.contains(tool_use_id.as_str()) {
            missing_by_index
                .entry(*assistant_idx)
                .or_default()
                .push(tool_use_id.clone());
        }
    }

    // Sort indices descending to insert without shifting issues
    let mut sorted_indices: Vec<usize> = missing_by_index.keys().copied().collect();
    sorted_indices.sort_unstable_by(|a, b| b.cmp(a));

    for assistant_idx in sorted_indices {
        if let Some(missing_ids) = missing_by_index.get(&assistant_idx) {
            // Insert in reverse to maintain original tool_use order
            for tool_call_id in missing_ids.iter().rev() {
                warn!(tool_call_id = %tool_call_id, "Injected synthetic tool result for interrupted execution");
                valid.insert(
                    assistant_idx + 1,
                    Message::ToolResult {
                        tool_call_id: tool_call_id.clone(),
                        content: ToolResultMessageContent::Text(INTERRUPTED_CONTENT.into()),
                        is_error: None,
                    },
                );
            }
        }
    }

    // PHASE 4: Ensure first message is user role
    if !valid.is_empty() && !valid[0].is_user() {
        warn!("Injected placeholder user message at start");
        valid.insert(0, Message::user(CONTINUED_CONTENT));
    }

    valid
}

/// Check if assistant content will survive API conversion.
///
/// Thinking blocks without signatures are filtered out before sending to the API.
/// If a message contains ONLY such blocks, it would become empty.
fn has_content_surviving_conversion(content: &[AssistantContent]) -> bool {
    content.iter().any(|block| match block {
        AssistantContent::Text { .. } | AssistantContent::ToolUse { .. } => true,
        AssistantContent::Thinking { signature, .. } => {
            signature.as_ref().is_some_and(|s| !s.is_empty())
        }
    })
}

/// Check if user message content is non-empty.
fn is_valid_user_content(content: &tron_core::messages::UserMessageContent) -> bool {
    match content {
        tron_core::messages::UserMessageContent::Text(text) => !text.trim().is_empty(),
        tron_core::messages::UserMessageContent::Blocks(blocks) => !blocks.is_empty(),
    }
}

/// Check if tool result content is non-empty.
fn is_valid_tool_result_content(content: &ToolResultMessageContent) -> bool {
    match content {
        ToolResultMessageContent::Text(_) => true,
        ToolResultMessageContent::Blocks(blocks) => !blocks.is_empty(),
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Tests
// ─────────────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Map;
    use tron_core::content::AssistantContent;
    use tron_core::messages::{Message, ToolResultMessageContent};

    fn tool_use(id: &str, name: &str) -> AssistantContent {
        AssistantContent::ToolUse {
            id: id.into(),
            name: name.into(),
            arguments: Map::new(),
            thought_signature: None,
        }
    }

    fn tool_result(tool_call_id: &str, text: &str) -> Message {
        Message::ToolResult {
            tool_call_id: tool_call_id.into(),
            content: ToolResultMessageContent::Text(text.into()),
            is_error: None,
        }
    }

    fn assistant_with_content(content: Vec<AssistantContent>) -> Message {
        Message::Assistant {
            content,
            usage: None,
            cost: None,
            stop_reason: None,
            thinking: None,
        }
    }

    // ── Phase 1: Filter invalid & deduplicate ────────────────────────────

    #[test]
    fn empty_messages_filtered() {
        let messages = vec![
            Message::user("hello"),
            Message::User {
                content: tron_core::messages::UserMessageContent::Text("  ".into()),
                timestamp: None,
            },
            Message::assistant("world"),
        ];
        let result = sanitize_messages(messages);
        assert_eq!(result.len(), 2);
        assert!(result[0].is_user());
        assert!(result[1].is_assistant());
    }

    #[test]
    fn thinking_only_assistant_filtered() {
        let messages = vec![
            Message::user("hello"),
            assistant_with_content(vec![AssistantContent::Thinking {
                thinking: "display only".into(),
                signature: None,
            }]),
            Message::assistant("visible"),
        ];
        let result = sanitize_messages(messages);
        assert_eq!(result.len(), 2);
        assert!(result[0].is_user());
        assert!(result[1].is_assistant());
    }

    #[test]
    fn thinking_with_signature_preserved() {
        let messages = vec![
            Message::user("hello"),
            assistant_with_content(vec![
                AssistantContent::Thinking {
                    thinking: "deep thought".into(),
                    signature: Some("sig123".into()),
                },
                AssistantContent::text("answer"),
            ]),
        ];
        let result = sanitize_messages(messages);
        assert_eq!(result.len(), 2);
    }

    #[test]
    fn duplicate_tool_use_ids_removed() {
        let messages = vec![
            Message::user("hello"),
            assistant_with_content(vec![
                tool_use("tc-1", "bash"),
                AssistantContent::text("some text"),
            ]),
            tool_result("tc-1", "output"),
            // Second assistant reuses the same tool_use ID
            assistant_with_content(vec![
                tool_use("tc-1", "bash"),
                AssistantContent::text("more text"),
            ]),
        ];
        let result = sanitize_messages(messages);
        // The second assistant message should have the duplicate tool_use removed
        // but still keep the text block
        assert_eq!(result.len(), 4);
        if let Message::Assistant { content, .. } = &result[3] {
            assert_eq!(content.len(), 1); // only text remains
            assert!(content[0].is_text());
        } else {
            panic!("Expected assistant message");
        }
    }

    // ── Phase 2-3: Synthetic tool results ────────────────────────────────

    #[test]
    fn unmatched_tool_use_gets_synthetic_result() {
        let messages = vec![
            Message::user("hello"),
            assistant_with_content(vec![tool_use("tc-1", "bash")]),
            // No tool result for tc-1
        ];
        let result = sanitize_messages(messages);
        assert_eq!(result.len(), 3);
        assert!(result[0].is_user());
        assert!(result[1].is_assistant());
        // Synthetic result injected
        if let Message::ToolResult {
            tool_call_id,
            content,
            ..
        } = &result[2]
        {
            assert_eq!(tool_call_id, "tc-1");
            if let ToolResultMessageContent::Text(text) = content {
                assert_eq!(text, "[Interrupted]");
            }
        } else {
            panic!("Expected synthetic tool result");
        }
    }

    #[test]
    fn matched_tool_use_no_synthetic_result() {
        let messages = vec![
            Message::user("hello"),
            assistant_with_content(vec![tool_use("tc-1", "bash")]),
            tool_result("tc-1", "output"),
        ];
        let result = sanitize_messages(messages);
        assert_eq!(result.len(), 3); // No synthetic result needed
    }

    #[test]
    fn multiple_unmatched_tool_uses_all_get_synthetic_results() {
        let messages = vec![
            Message::user("hello"),
            assistant_with_content(vec![
                tool_use("tc-1", "bash"),
                tool_use("tc-2", "read"),
                tool_use("tc-3", "write"),
            ]),
        ];
        let result = sanitize_messages(messages);
        // user + assistant + 3 synthetic results
        assert_eq!(result.len(), 5);
        assert!(result[2].is_tool_result());
        assert!(result[3].is_tool_result());
        assert!(result[4].is_tool_result());
    }

    // ── Phase 4: First message must be user ──────────────────────────────

    #[test]
    fn first_message_assistant_gets_user_prepended() {
        let messages = vec![assistant_with_content(vec![AssistantContent::text(
            "response",
        )])];
        let result = sanitize_messages(messages);
        assert_eq!(result.len(), 2);
        assert!(result[0].is_user());
        if let Message::User {
            content: tron_core::messages::UserMessageContent::Text(text),
            ..
        } = &result[0]
        {
            assert_eq!(text, "[Continued]");
        }
    }

    #[test]
    fn first_message_user_unchanged() {
        let messages = vec![Message::user("hello"), Message::assistant("world")];
        let result = sanitize_messages(messages);
        assert_eq!(result.len(), 2);
        assert!(result[0].is_user());
    }

    // ── Idempotent ───────────────────────────────────────────────────────

    #[test]
    fn clean_messages_unchanged() {
        let messages = vec![
            Message::user("hello"),
            Message::assistant("world"),
            Message::user("follow-up"),
        ];
        let result = sanitize_messages(messages.clone());
        assert_eq!(result.len(), 3);
    }

    #[test]
    fn idempotent_double_sanitize() {
        let messages = vec![
            assistant_with_content(vec![tool_use("tc-1", "bash")]),
            Message::user("hello"),
        ];
        let first = sanitize_messages(messages);
        let second = sanitize_messages(first.clone());
        assert_eq!(first.len(), second.len());
    }

    // ── Multiple consecutive ToolResults preserved ───────────────────────

    #[test]
    fn multiple_consecutive_tool_results_preserved() {
        let messages = vec![
            Message::user("hello"),
            assistant_with_content(vec![
                tool_use("tc-1", "bash"),
                tool_use("tc-2", "read"),
                tool_use("tc-3", "write"),
            ]),
            tool_result("tc-1", "output1"),
            tool_result("tc-2", "output2"),
            tool_result("tc-3", "output3"),
        ];
        let result = sanitize_messages(messages);
        // All messages should be preserved as-is
        assert_eq!(result.len(), 5);
    }

    #[test]
    fn empty_input_returns_empty() {
        let result = sanitize_messages(vec![]);
        assert!(result.is_empty());
    }
}
